nilearn/surface.py

```diff
@@ -0,0 +1,572 @@
+
+    """
+    vertices, faces = mesh
+    offsets_world_space = _uniform_ball_cloud(
+        dim=vertices.shape[1], n_points=n_points) * ball_radius
+    mesh_voxel_space = np.asarray(
+        resampling.coord_transform(*vertices.T,
+                                   affine=np.linalg.inv(affine))).T
+    linear_map = np.eye(affine.shape[0])
+    linear_map[:-1, :-1] = affine[:-1, :-1]
+    offsets_voxel_space = np.asarray(
+        resampling.coord_transform(*offsets_world_space.T,
+                                   affine=np.linalg.inv(linear_map))).T
+    sample_locations_voxel_space = (mesh_voxel_space[:, np.newaxis, :] +
+                                    offsets_voxel_space[np.newaxis, :])
+    return sample_locations_voxel_space
+
+
+def _line_sample_locations(
+        mesh, affine, segment_half_width=3., n_points=10):
+    """Locations to draw samples from to project volume data onto a mesh.
+
+    For each mesh vertex, the locations of `n_points` points evenly spread in a
+    segment of the normal to the vertex are returned. The line segment has
+    length 2 * `segment_half_width` and is centered at the vertex.
+
+    Parameters
+    ----------
+    mesh : pair of np arrays.
+        mesh[0] contains the 3d coordinates of the vertices
+        (shape n_vertices, 3)
+        mesh[1] contains, for each triangle, the indices into mesh[0] of its
+        vertices (shape n_triangles, 3)
+
+    affine : array of shape (4, 4)
+        affine transformation from image voxels to the vertices' coordinate
+        space.
+
+    segment_half_width : float, optional (default=3.)
+        size in mm of the neighbourhood around each vertex in which to draw
+        samples
+
+    n_points : int, optional (default=10)
+        number of samples to draw for each vertex.
+
+    Returns
+    -------
+    numpy array, shape (n_vertices, n_points, 3)
+        The locations, in voxel space, from which to draw samples.
+        First dimension iterates over mesh vertices, second dimension iterates
+        over the sample points associated to a vertex, third dimension is x, y,
+        z in voxel space.
+
+    """
+    vertices, faces = mesh
+    normals = _vertex_outer_normals(mesh)
+    offsets = np.linspace(-segment_half_width, segment_half_width, n_points)
+    sample_locations = vertices[
+        np.newaxis, :, :] + normals * offsets[:, np.newaxis, np.newaxis]
+    sample_locations = np.rollaxis(sample_locations, 1)
+    sample_locations_voxel_space = np.asarray(
+        resampling.coord_transform(
+            *np.vstack(sample_locations).T,
+            affine=np.linalg.inv(affine))).T.reshape(sample_locations.shape)
+    return sample_locations_voxel_space
+
+
+def _sample_locations(mesh, affine, radius, kind='line', n_points=None):
+    """Get either ball or line sample locations."""
+    projectors = {
+        'line': _line_sample_locations,
+        'ball': _ball_sample_locations
+    }
+    if kind not in projectors:
```

Comment:
Could you please also check if n_samples and radius is ```import numbers.Number``` ? AFAIK there is no validation on this ?

Reply:
Is this something that we usually do? In other nilearn functions I don't see this kind of type checking

Reply:
Just to make it much user friendly if accidentally got some weird inputs. I think it is a minor comment which you can avoid.